sparse/unit_test/Test_Sparse_removeCrsMatrixZeros.hpp

```diff
@@ -0,0 +1,193 @@
+      // No entries, but nonzero dimensions.
+      std::vector<int> rowmap = {0, 0, 0, 0, 0};
+      std::vector<int> entries;
+      std::vector<double> values;
+      return loadMatrixFromVectors<Matrix>(4, 4, rowmap, entries, values);
+    }
+    case 1: {
+      // Some empty rows, and some zero values
+      std::vector<int> rowmap    = {0, 0, 3, 3, 5};
+      std::vector<int> entries   = {0, 1, 3, 1, 2};
+      std::vector<double> values = {1, 3, 0, 0, 2};
+      return loadMatrixFromVectors<Matrix>(4, 4, rowmap, entries, values);
+    }
+    case 2: {
+      // Zero-row matrix, length-0 rowmap
+      typename Matrix::row_map_type rowmap;
+      typename Matrix::index_type entries;
+      typename Matrix::values_type values;
+      return Matrix("A empty", 0, 0, 0, values, rowmap, entries);
+    }
+    case 3: {
+      // Zero-row matrix, length-1 rowmap
+      std::vector<int> rowmap = {0};
+      std::vector<int> entries;
+      std::vector<double> values;
+      return loadMatrixFromVectors<Matrix>(0, 0, rowmap, entries, values);
+    }
+    case 4: {
+      // A row of all zeros that will be filtered
+      std::vector<int> rowmap    = {0, 3, 6};
+      std::vector<int> entries   = {0, 1, 2, 3, 4, 5};
+      std::vector<double> values = {0, 0, 0, 1, 1, 1};
+      return loadMatrixFromVectors<Matrix>(2, 6, rowmap, entries, values);
+    }
+    case 5: {
+      // One zero in each row that will be filtered
+      std::vector<int> rowmap    = {0, 2, 4, 7};
+      std::vector<int> entries   = {0, 1, 1, 2, 0, 1, 2};
+      std::vector<double> values = {0, 1, 1, 0, 0, 3, -3};
+      return loadMatrixFromVectors<Matrix>(3, 3, rowmap, entries, values);
+    }
+    case 6: {
+      // First and last rows empty
+      std::vector<int> rowmap    = {0, 0, 2, 2};
+      std::vector<int> entries   = {0, 1};
+      std::vector<double> values = {0, 3.14};
+      return loadMatrixFromVectors<Matrix>(3, 2, rowmap, entries, values);
+    }
+    case 7: {
+      // First and last rows nonempty, but will be empty after filtering
+      std::vector<int> rowmap    = {0, 2, 4, 6};
+      std::vector<int> entries   = {0, 1, 1, 2, 0, 3};
+      std::vector<double> values = {0, 0, 1, -1, 0, 0};
+      return loadMatrixFromVectors<Matrix>(3, 4, rowmap, entries, values);
+    }
+    case 8: {
+      // Large, random matrix with 30% of values converted to zero
+      Offset nnz = 40 * 10000;
+      Matrix A   = KokkosSparse::Impl::kk_generate_sparse_matrix<Matrix>(
+          10000, 10000, nnz, 10, 5000);
+      auto valuesHost = Kokkos::create_mirror_view(A.values);
+      Kokkos::deep_copy(valuesHost, A.values);
+      for (Offset i = 0; i < A.nnz(); i++) {
+        if (rand() % 10 < 3) valuesHost(i) = 0.0;
+      }
+      Kokkos::deep_copy(A.values, valuesHost);
+      return A;
+    }
+  }
+  throw std::invalid_argument("Test case number of out bounds");
+}
+
+}  // namespace TestRemoveCrsMatrixZeros
+
+void testRemoveCrsMatrixZeros(int testCase) {
```

Comment:
Since the matrices on which you perform the tests are hard coded, you could use one test case to check that `removeMatrixZerosReference()` returns the expected answers.
Ideally doing that for all hard-coded cases is best. Using the reference routines can then be employed with randomly generated inputs with confidence.